botorch/models/transforms/input.py

```diff
@@ -458,6 +458,136 @@ def equals(self, other: InputTransform) -> bool:
         return False
 
 
+class Standardize(ReversibleInputTransform, Module):
+    r"""Standardize inputs (zero mean, unit variance).
+
+    In train mode, calling `forward` updates the module state
+    (i.e. the mean/std normalizing constants). If in eval mode, calling `forward`
+    simply applies the standardization using the current module state.
+    """
+
+    def __init__(
+        self,
+        d: int,
+        indices: Optional[List[int]] = None,
+        batch_shape: torch.Size = torch.Size(),  # noqa: B008
+        transform_on_train: bool = True,
+        transform_on_eval: bool = True,
+        transform_on_fantasize: bool = True,
+        reverse: bool = False,
+        min_std: float = 1e-8,
+    ) -> None:
+        r"""Standardize inputs (zero mean, unit variance).
+
+        Args:
+            d: The dimension of the input space.
+            indices: The indices of the inputs to standardize. If omitted,
+                take all dimensions of the inputs into account.
+            batch_shape: The batch shape of the inputs (asssuming input tensors
+                of shape `batch_shape x n x d`). If provided, perform individual
+                normalization per batch, otherwise uses a single normalization.
+            transform_on_train: A boolean indicating whether to apply the
+                transforms in train() mode. Default: True
+            transform_on_eval: A boolean indicating whether to apply the
+                transform in eval() mode. Default: True
+            reverse: A boolean indicating whether the forward pass should untransform
+                the inputs.
+            min_std: Amount of noise to add to the standard deviation to ensure no
+                division by zero errors.
+        """
+        super().__init__()
+        if (indices is not None) and (len(indices) == 0):
+            raise ValueError("`indices` list is empty!")
+        if (indices is not None) and (len(indices) > 0):
+            indices = torch.tensor(indices, dtype=torch.long)
+            if len(indices) > d:
+                raise ValueError("Can provide at most `d` indices!")
+            if (indices > d - 1).any():
+                raise ValueError("Elements of `indices` have to be smaller than `d`!")
+            if len(indices.unique()) != len(indices):
+                raise ValueError("Elements of `indices` tensor must be unique!")
+            self.indices = indices
```

Comment:
I also thought about it because from a purely mathematical point of view `Standardize` and `Normalize` are almost the same, both are linear transformations of the form: `X_transformed = (X-subtrahend)/divisor`.

Only difference is that its coefficients (`subtrahend ` and `divisor`) are calculated differently. A possible solution would be introducing an additional class called `LinearTransform` and inherit then from this. This could look like this:

```python
class LinearTransform(ReversibleInputTransform, Module):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        coefficients: Optional(Tuple(Tensor,Tensor)) = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_divisor: float = 1e-8,
    ) -> None:
        super().__init__()
        if (indices is not None) and (len(indices) == 0):
            raise ValueError("`indices` list is empty!")
        if (indices is not None) and (len(indices) > 0):
            indices = torch.tensor(indices, dtype=torch.long)
            if len(indices) > d:
                raise ValueError("Can provide at most `d` indices!")
            if (indices > d - 1).any():
                raise ValueError("Elements of `indices` have to be smaller than `d`!")
            if len(indices.unique()) != len(indices):
                raise ValueError("Elements of `indices` tensor must be unique!")
            self.indices = indices
        if coefficients is not None:
            if coefficients[0].size(-1) != d:
                raise BotorchTensorDimensionError(
                    "Dimensions of provided subtrahend are incompatible with `d`!"
                )

            if coefficients[1].size(-1) != d:
                raise BotorchTensorDimensionError(
                    "Dimensions of provided divisor are incompatible with `d`!"
                )
            subtrahend = self.coefficients[0]
            divisor = self.coefficients[1]
            self.learn_coefficients = False
        else:
            subtrahend = torch.zeros(*batch_shape, 1, d)
            divisor = torch.zeros(*batch_shape, 1, d)
            self.learn_bounds = True
        self.register_buffer("subtrahend", subtrahend)
        self.register_buffer("divisor", divisor)
        self._d = d
        self.transform_on_train = transform_on_train
        self.transform_on_eval = transform_on_eval
        self.transform_on_fantasize = transform_on_fantasize
        self.reverse = reverse
        self.batch_shape = batch_shape
        self.min_divisor = min_divisor

    def _transform(self, X: Tensor) -> Tensor:
        if self.learn_coefficients and self.training:
            if X.size(-1) != self.divisor.size(-1):
                raise BotorchTensorDimensionError(
                    f"Wrong input dimension. Received {X.size(-1)}, "
                    f"expected {self.divisor.size(-1)}."
                )
            self.subtrahend, self.divisor = self._learn_coefficients()
            self.divisor[torch.where(self.divisor <= self.min_divisor)] = self.min_divisor
        
        if hasattr(self, "indices"):
            X_new = X.clone()
            X_new[..., self.indices] = (
                X_new[..., self.indices] - self.subtrahend[..., self.indices]
            ) / self.divisor[..., self.indices]
            return X_new
        return (X - self.subtrahend) / self.divisor

    def _untransform(self, X: Tensor) -> Tensor:
        if hasattr(self, "indices"):
            X_new = X.clone()
            X_new[..., self.indices] = (
                self.subtrahend[..., self.indices]
                + X_new[..., self.indices] * self.divisor[..., self.indices]
            )
            return X_new
        return self.subtrahend + X * self.divisor

    @abstractmethod
    def learn_coefficients(self):
        return subtrahend, divisor


class Normalize(LinearTransform):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        bounds: Optional[Tensor] = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_range: float = 1e-8,
    ) -> None:
        if bounds is not None:
            subtrahend = bounds[..., 0:1, :]
            divisor = bounds[..., 1:2, :] - subtrahend
        super().__init__(
            d=d,
            indices=indices,
            coefficients=(subtrahend,divisor),
            batch_shape=batch_shape,
            transform_on_train=transform_on_train,
            transform_on_eval=transform_on_eval,
            transform_on_fantasize=transform_on_fantasize,
            reverse=reverse,
            min_divisor=min_range)
        return

    def learn_coefficients(self):
        mins = X.min(dim=-2, keepdim=True)[0]
        ranges = X.max(dim=-2, keepdim=True)[0] - mins
        return mins, ranges


class InputStandardize(LinearTransform):

    def __init__(
        self,
        d: int,
        indices: Optional[List[int]] = None,
        batch_shape: torch.Size = torch.Size(),  # noqa: B008
        transform_on_train: bool = True,
        transform_on_eval: bool = True,
        transform_on_fantasize: bool = True,
        reverse: bool = False,
        min_std: float = 1e-8,
    ) -> None:
        super().__init__(
            d=d,
            indices=indices,
            coefficients=None,
            batch_shape=batch_shape,
            transform_on_train=transform_on_train,
            transform_on_eval=transform_on_eval,
            transform_on_fantasize=transform_on_fantasize,
            reverse=reverse,
            min_divisor=min_std)
        return

    def learn_coefficients(self):
        means = X.mean(dim=-2, keepdim=True)
        stds = X.std(dim=-2, keepdim=True)
        return means, stds
```

Note this code is not tested or properly formatted and should only solve as an example! Maybe one could integrate something like this in a separate PR as you suggested.

Reply:
yeah it's a nice idea to factor out that functionality and consolidate it into a common base class like this. Let's do this in a separate PR. Would you be willing to submit one?